### PR TITLE
Removes try/catch from context

### PR DESCRIPTION
**Pull request by badisa**
* Uses DeviceBuffers instead

### CODE REVIEW

File: timemachine/cpp/src/context.cu [hunk 1]
```diff
@@ -74,65 +61,65 @@ Context::multiple_steps(const std::vector<double> &lambda_schedule, int store_du
...
-            }
+    std::unique_ptr<DeviceBuffer<double>> d_box_buffer(nullptr);
+    if (box_buffer_size > 0) {
+        d_box_buffer.reset(new DeviceBuffer<double>(box_buffer_size));
+    }
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dl_buffer(nullptr);
+    if (du_dl_buffer_size > 0) {
+        d_du_dl_buffer.reset(new DeviceBuffer<unsigned long long>(du_dl_buffer_size));
+    }
+    gpuErrchk(cudaMemsetAsync(d_du_dl_buffer->data, 0, d_du_dl_buffer->size, stream));
```
Comment (badisa) [inline timemachine/cpp/src/context.cu:75]: This is wrong

File: timemachine/cpp/src/context.cu [hunk 2]
```diff
@@ -74,65 +61,65 @@ Context::multiple_steps(const std::vector<double> &lambda_schedule, int store_du
...
-            }
+    std::unique_ptr<DeviceBuffer<double>> d_box_buffer(nullptr);
+    if (box_buffer_size > 0) {
+        d_box_buffer.reset(new DeviceBuffer<double>(box_buffer_size));
+    }
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dl_buffer(nullptr);
+    if (du_dl_buffer_size > 0) {
+        d_du_dl_buffer.reset(new DeviceBuffer<unsigned long long>(du_dl_buffer_size));
+        gpuErrchk(cudaMemsetAsync(d_du_dl_buffer->data, 0, d_du_dl_buffer->size, stream));
+    }
```
Comment (mcwitt) [inline timemachine/cpp/src/context.cu:75]: Just noticing that we could avoid `unique_ptr` and dynamic allocation here except that `DeviceBuffer` currently has a check that fails if the size is zero. Would it be worth relaxing that restriction? (conflicts with e.g. the behavior of `malloc` AFAIK)

Comment (badisa) [reply timemachine/cpp/src/context.cu:75]: I would be for removing the check as it is consistent with cudaMalloc's behavior (nullptr if size is 0). Last time this came up though we decided against it, though I can't find a reference to that conversation.

Easily done in this PR and would simplify a fair amount of code.

Comment (mcwitt) [reply timemachine/cpp/src/context.cu:75]: To be clear, I wouldn't argue for using a literal `DeviceBuffer(0)` in cases where we don't want to allocate (I think the discussion starting [here](https://github.com/proteneer/timemachine/pull/726#discussion_r853317229) was more about that case)

**Review by mcwitt — APPROVED** [review]
LGTM. Nice cleanup!

**Comment by proteneer** [comment]
hopefully once @mcwitt finishes the du_dl item and we remove lamb from the signature of U, we can completely remove d_du_dl  from these context methods!

File: timemachine/cpp/src/context.cu (hunk 2 above)
Comment (proteneer) [reply timemachine/cpp/src/context.cu:75]: 1) I agree that wrapping a `DeviceBuffer` in a `unique_ptr` is unwieldy.
2) Do we actually have any practical situations where `box_buffer_size`, and `du_dl_buffer_size` is zero? If not, I'd just turn assert that those two quantities are `> 0` and simplify.

**Review by proteneer — APPROVED** [review]

**Pull request merged**